rclcpp_action/include/rclcpp_action/server.hpp

```diff
@@ -369,7 +369,11 @@ class Server : public ServerBase, public std::enable_shared_from_this<Server<Act
     if (goal_handle) {
       resp = handle_cancel_(goal_handle);
       if (CancelResponse::ACCEPT == resp) {
-        goal_handle->_cancel_goal();
+        try {
+          goal_handle->_cancel_goal();
+        } catch (const rclcpp::exceptions::RCLError & ex) {
```

Comment:
According to this file https://github.com/ros2/rclcpp/blob/master/rclcpp_action/src/server_goal_handle.cpp, I see that all errors of transiting state failed are throwing `RCLError` exception. Is it better to regard them as the same error? What do you think?

Reply:
> And if you are talking about adding a new exception to clarify this type of error. I am not sure if it would affect users‘ implementation of rclcpp_action because they should catch another exception?

Yes, I meant that.
`rcl_action_update_goal_state` has two return values, one is the one [we want to catch](https://github.com/ros2/rcl/blob/414fdbd479b83f2b8af8235ad1fa3aeba2b0027f/rcl_action/src/rcl_action/goal_handle.c#L99), the [other one](https://github.com/ros2/rcl/blob/414fdbd479b83f2b8af8235ad1fa3aeba2b0027f/rcl_action/src/rcl_action/goal_handle.c#L90) shouldn't be catched.

It would be good to have a more specific exception for the one we want to catch.
@jacobperron what do you think?

Reply:
I agree that it would be nicer to have more specific exceptions. I also agree that we shouldn't expect a null goal handle. However, it may not be API compatible to change exception types, so I suggest we do that for H-turtle. If we leave this PR as it is, then we can backport this change to Galactic (and maybe Foxy).

Reply:
we are already doing similar stuff in other places (e.g. [here](https://github.com/ros2/rclcpp/blob/1037822a63330495dcf0de5e8f20544375a5f116/rclcpp_action/src/server_goal_handle.cpp#L120-L122)), so I'm ok with it.

> However, it may not be API compatible to change exception types

If the new exception is a subclass of `RCLError`, it will be both API and ABI compatible.